man/dmsend.1
.TH "DMSEND" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmsend \(en sends observations and logs to POSIX message queues
.SH SYNOPSIS
\f[B]dmsend\f[R] \-\-help
.PP
\f[B]dmsend\f[R] \-\-version
.PP
\f[B]dmsend\f[R] \-\-\f[B]type\f[R] [log|observ] \-\-\f[B]format\f[R]
[csv|nml] \-\-\f[B]input\f[R] \f[I]file\f[R] [\-\-\f[B]receiver\f[R]
\f[I]name\f[R]] [\-\-\f[B]name\f[R] \f[I]name\f[R]] [\-\-\f[B]node\f[R]
\f[I]id\f[R]] [\-\-\f[B]logger\f[R] \f[I]name\f[R]]
[\-\-\f[B]debug\f[R]] [\-\-\f[B]forward\f[R]] [\-\-\f[B]verbose\f[R]]
.PP
\f[B]dmsend\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R] [\-\-\f[B]name\f[R]
\f[I]name\f[R]]
.SH DESCRIPTION
The \f[B]dmsend\f[R] program reads observations or logs in CSV or
Fortran 95 Namelist format, and sends them sequentially to the POSIX
message queue of the given receiver.
The data is either read from file or from standard input.
If the input data is of type \f[CR]observ\f[R] and the argument
\f[CR]\-\-forward\f[R] is passed, each observation will be sent to its
next specified receiver in the receivers list.
If no receivers are declared, or if the end of the receivers list is
reached, the observation will not be forwarded.
.PP
The program settings are passed through command\-line arguments or an
optional configuration file.
The arguments overwrite settings from file.
.SH OPTIONS
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
Path to configuration file.
.TP
\f[B]\-\-debug\f[R], \f[B]\-D\f[R]
Forward logs messages of level \f[CR]LL_DEBUG\f[R] via IPC (if logger is
set).
.TP
\f[B]\-\-format\f[R], \f[B]\-f\f[R] [csv|nml]
Input format, either CSV or Fortran 95 Namelist.
.TP
\f[B]\-\-forward\f[R], \f[B]\-F\f[R]
Forward observations to the next specified receiver in the receivers
list.
Not allowed in combination with argument \f[CR]\-\-receiver\f[R].
.TP
\f[B]\-\-logger\f[R], \f[B]\-l\f[R] \f[I]name\f[R]
Name of logger.
If set, sends logs to \f[I]dmlogger(1)\f[R] process of given name.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-input\f[R], \f[B]\-i\f[R]
Path to input file, either in CSV or Namelist format.
Reads records from \f[I]stdin\f[R] if not passed or set to
\f[CR]\-\f[R].
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of program instance and configuration (default is
\f[CR]dmsend\f[R]).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Optional node id.
.TP
\f[B]\-\-receiver\f[R], \f[B]\-r\f[R] \f[I]name\f[R]
Name of the observation or log receiver, without leading \f[CR]/\f[R].
Not allowed in combination with argument \f[CR]\-\-forward\f[R].
.TP
\f[B]\-\-type\f[R], \f[B]\-t\f[R] [log|observ]
Data type to send, either log or observation.
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print log messages to \f[I]stderr\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Program terminated normally.
.TP
\f[B]1\f[R]
Failure.
Program execution failed.
.SH ENVIRONMENT
.TP
\f[B]DM_LOGGER\f[R]
Name of logger instance to send logs to.
.TP
\f[B]NO_COLOR\f[R]
Disable ANSI colour output.
.SH EXAMPLE
Read observation from Namelist file \f[CR]observ.nml\f[R] and send it to
the next specified receiver:
.IP
.EX
$ dmsend \-t observ \-f nml \-i observ.nml \-F
.EE
.PP
Send logs in CSV file \f[CR]logs.csv\f[R] sequentially to process
\f[CR]dmrecv\f[R]:
.IP
.EX
$ dmsend \-r dmrecv \-t log \-f csv \-i logs.csv
.EE
.SH SEE ALSO
\f[I]dmrecv(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmsend.f90
! dmsend.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmsend
    !! Reads logs or observations in CSV or Fortran 95 Namelist format from
    !! file or standard input, then sends all records to a POSIX message queue.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmsend'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    logical, parameter :: APP_MQ_BLOCKING = .true. !! Observation forwarding is blocking.

    type :: app_type
        !! Application settings.
        character(ID_LEN)              :: name        = APP_NAME    !! Name of process and POSIX message queue.
        character(FILE_PATH_LEN)       :: config      = ' '         !! Path to configuration file.
        character(LOGGER_NAME_LEN)     :: logger      = ' '         !! Name of logger (name implies IPC).
        character(NODE_ID_LEN)         :: node_id     = ' '         !! Optional node id.
        character(FILE_PATH_LEN)       :: input       = ' '         !! Path to input file (stdin if empty or `-`).
        character(FORMAT_NAME_LEN)     :: format_name = ' '         !! Format name.
        character(TYPE_NAME_LEN)       :: type_name   = ' '         !! Type name.
        character(OBSERV_RECEIVER_LEN) :: receiver    = ' '         !! Name of receiver's message queue (without leading `/`).
        integer                        :: format      = FORMAT_NONE !! Input format.
        integer                        :: type        = TYPE_NONE   !! Data type.
        logical                        :: debug       = .false.     !! Forward debug messages via IPC.
        logical                        :: forward     = .false.     !! Enable observation forwarding.
        logical                        :: verbose     = .false.     !! Print debug messages to stderr.
    end type app_type

    class(logger_class), pointer :: logger ! Logger object.

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App settings.

    ! Initialise DMPACK.
    call dm_init()

    ! Read command-line arguments and options from configuration file.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Initialise logger.
    logger => dm_logger_get_default()
    call logger%configure(name    = app%logger,  & ! Name of logger process.
                          node_id = app%node_id, & ! Node id.
                          source  = app%name,    & ! Log source.
                          debug   = app%debug,   & ! Forward debug messages via IPC.
                          ipc     = .true.,      & ! Enable IPC (if logger is set).
                          verbose = app%verbose)   ! Print logs to standard error.
    call logger%info('started ' // APP_NAME)

    rc = run(app)
    call shutdown(rc)
contains
    integer function run(app) result(rc)
        !! Reads logs or observations from file/standard input, and then sends
        !! all records to one or more POSIX messages queues.
        type(app_type), intent(inout) :: app !! App type.

        integer                 :: file_unit, stat
        integer(i8)             :: nrecords
        logical                 :: is_file
        type(log_type)          :: log
        type(observ_type)       :: observ
        type(posix_mqueue_type) :: mqueue

        nrecords  = 0
        file_unit = stdin
        is_file   = .false.

        if (dm_string_has(app%input) .and. app%input /= '-') is_file = .true.

        ! Open message queue of receiver for writing.
        if (.not. app%forward) then
            call logger%debug('opening mqueue /' // app%receiver)
            rc = dm_posix_mqueue_open(mqueue, type=app%type, name=app%receiver, access=POSIX_MQUEUE_WRONLY, blocking=.true.)

            if (dm_is_error(rc)) then
                call logger%error('failed to open mqueue /' // app%receiver, error=rc)
                return
            end if
        end if

        read_block: block
            if (is_file) then
                ! Open input file.
                call logger%debug('opening input file ' // app%input)
                open (action='read', file=trim(app%input), iostat=stat, newunit=file_unit, status='old')

                if (stat /= 0) then
                    rc = E_IO
                    call logger%error('failed to open input file ' // app%input, error=rc)
                    exit read_block
                end if
            end if

            ! Read serialised log or observation from file/stdin.
            ipc_loop: do
                ! **************************************************************
                ! OBSERVATION TYPE.
                ! **************************************************************
                if (app%type == TYPE_OBSERV) then
                   ! Read observation in CSV or Namelist format.
                    select case (app%format)
                        case (FORMAT_CSV); rc = dm_csv_read(observ, unit=file_unit)
                        case (FORMAT_NML); rc = dm_nml_read(observ, unit=file_unit)
                    end select

                    ! End of file reached.
                    if (rc == E_EOF) then
                        rc = E_NONE
                        call logger%debug('end of file reached')
                        exit ipc_loop
                    end if

                    if (dm_is_error(rc)) then
                        call logger%error('failed to read observation', error=rc)
                        exit ipc_loop
                    end if

                    select case (app%format)
                        case (FORMAT_CSV); call logger%debug('read observation in CSV format')
                        case (FORMAT_NML); call logger%debug('read observation in NML format')
                    end select

                    ! Validate input.
                    if (.not. dm_observ_is_valid(observ)) then
                        call logger%debug('invalid input observation ' // observ%id, error=E_INVALID)
                    end if

                    ! Forward observation to next receiver, or send it to message queue.
                    if (app%forward) then
                        rc = dm_posix_mqueue_forward(observ, name=app%name, blocking=APP_MQ_BLOCKING)
                    else
                        rc = dm_posix_mqueue_write(mqueue, observ)
                    end if
                ! **************************************************************
                ! LOG TYPE.
                ! **************************************************************
                else if (app%type == TYPE_LOG) then
                    ! Read log in CSV or Namelist format.
                    select case (app%format)
                        case (FORMAT_CSV); rc = dm_csv_read(log, unit=file_unit)
                        case (FORMAT_NML); rc = dm_nml_read(log, unit=file_unit)
                    end select

                    ! End of file reached.
                    if (rc == E_EOF) then
                        rc = E_NONE
                        call logger%debug('end of file reached')
                        exit ipc_loop
                    end if

                    if (dm_is_error(rc)) then
                        call logger%error('failed to read log', error=rc)
                        exit ipc_loop
                    end if

                    select case (app%format)
                        case (FORMAT_CSV); call logger%debug('read log in CSV format')
                        case (FORMAT_NML); call logger%debug('read log in NML format')
                    end select

                    ! Validate input.
                    if (.not. dm_log_is_valid(log)) then
                        call logger%error('invalid input log ' // log%id, error=E_INVALID)
                        cycle ipc_loop
                    end if

                    ! Send log to message queue.
                    rc = dm_posix_mqueue_write(mqueue, log)
                end if

                ! Handle message queue error.
                if (dm_is_error(rc)) then
                    call logger%error('failed to write to mqueue', error=rc)
                    call dm_sleep(1)
                    cycle ipc_loop
                end if

                nrecords = max(0_i8, nrecords + 1)
            end do ipc_loop

            if (is_file) then
                call logger%debug('closing input file ' // app%input)
                close (file_unit)
            end if
        end block read_block

        ! Close message queue.
        if (.not. app%forward) then
            call dm_posix_mqueue_close(mqueue, error=stat)
            call logger%debug('closed mqueue /' // app%receiver, error=stat)
        end if

        call logger%debug('finished transmission of ' // dm_itoa(nrecords) // ' records')
    end function run

    subroutine shutdown(error)
        !! Stops program.
        integer, intent(in) :: error !! DMPACK error code.

        integer :: stat

        stat = merge(STOP_FAILURE, STOP_SUCCESS, dm_is_error(error))
        call logger%info('stopped ' // APP_NAME, error=error)
        call dm_stop(stat)
    end subroutine shutdown

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from configuration file.
        type(app_type), intent(out) :: app

        type(arg_parser_class) :: parser

        ! Required and optional command-line arguments.
        call parser%add('name',     short='n', type=ARG_TYPE_ID)      ! -n, --name <string>
        call parser%add('config',   short='c', type=ARG_TYPE_FILE)    ! -c, --config <path>
        call parser%add('logger',   short='l', type=ARG_TYPE_ID)      ! -l, --logger <string>
        call parser%add('node',     short='N', type=ARG_TYPE_ID)      ! -N, --node <string>
        call parser%add('input',    short='i', type=ARG_TYPE_FILE)    ! -i, --input <path>
        call parser%add('format',   short='f', type=ARG_TYPE_STRING)  ! -f, --format <string>
        call parser%add('type',     short='t', type=ARG_TYPE_STRING)  ! -t, --type <string>
        call parser%add('receiver', short='r', type=ARG_TYPE_ID, max_len=OBSERV_RECEIVER_LEN) ! -r, --receiver <string>
        call parser%add('debug',    short='D', type=ARG_TYPE_LOGICAL) ! -D, --debug
        call parser%add('forward',  short='F', type=ARG_TYPE_LOGICAL) ! -F, --forward
        call parser%add('verbose',  short='V', type=ARG_TYPE_LOGICAL) ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Overwrite settings.
        call parser%get('logger',   app%logger)
        call parser%get('node',     app%node_id)
        call parser%get('input',    app%input)
        call parser%get('format',   app%format_name)
        call parser%get('type',     app%type_name)
        call parser%get('receiver', app%receiver)
        call parser%get('debug',    app%debug)
        call parser%get('forward',  app%forward)
        call parser%get('verbose',  app%verbose)

        app%format = dm_format_from_name(app%format_name)
        app%type   = dm_type_from_name(app%type_name)

        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads configuration from (Lua) file if path is not emty.
        type(app_type), intent(inout) :: app !! App type.
        type(config_class)             :: config

        rc = E_NONE
        if (.not. dm_string_has(app%config)) return

        rc = config%open(app%config, app%name)

        if (dm_is_ok(rc)) then
            call config%get('logger',   app%logger)
            call config%get('node',     app%node_id)
            call config%get('input',    app%input)
            call config%get('format',   app%format_name)
            call config%get('receiver', app%receiver)
            call config%get('type',     app%type_name)
            call config%get('debug',    app%debug)
            call config%get('forward',  app%forward)
            call config%get('verbose',  app%verbose)
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (.not. dm_id_is_valid(app%name)) then
            call dm_error_out(rc, 'invalid name')
            return
        end if

        if (dm_string_has(app%node_id) .and. .not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid node id')
            return
        end if

        if (dm_string_has(app%logger) .and. .not. dm_id_is_valid(app%logger)) then
            call dm_error_out(rc, 'invalid logger name')
            return
        end if

        if (app%type /= TYPE_OBSERV .and. app%type /= TYPE_LOG) then
            call dm_error_out(rc, 'invalid type')
            return
        end if

        if (app%format /= FORMAT_CSV .and. app%format /= FORMAT_NML) then
            call dm_error_out(rc, 'invalid format')
            return
        end if

        if (app%forward .and. app%type /= TYPE_OBSERV) then
            call dm_error_out(rc, '--forward requires type observation')
            return
        else if (.not. app%forward .and. .not. dm_id_is_valid(app%receiver)) then
            call dm_error_out(rc, 'invalid receiver')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_lua_version(.true.)
    end subroutine version_callback
end program dmsend
